File: data_structures/dimaeter_of_a_tree/cpp/Diameter.cpp
//WAP to find diameter of a BST.
#include<iostream>
#include<malloc.h>
using namespace std;
struct tree{
	int data;
	struct tree *left,*right;
};
typedef struct tree node;
node *current=NULL,*root=NULL,*parent=NULL,*temp=NULL;
void newnode(int n)
{
	current = (node*)malloc(sizeof(node));
	current->data = n;
	current->left = NULL;
	current->right = NULL;
	if(root==NULL)
	{
		root=current;
	}
	else
	{
		temp=root;
		parent=NULL;
		while(1)
		{	parent = temp;
			if(n<temp->data)
			{
				temp = temp->left;
				if(temp==NULL)
				{	parent->left= current;
					return;
				}
			}
			else
			{
				temp = temp->right;
				if(temp==NULL)
				{	parent->right= current;
					return;
				}
			}	
		}
	}
}
int max(int a,int b)
{
	if(a>b)
	return a;
	else
	return b;
}
int height(node* temp)
{
	if(temp==NULL)
	return 0;
	return 1+max(height(temp->left),height(temp->right));}
int diameter(node* temp)
{
	int dia=0,maxDia;
	if(temp ==NULL)
	return 0;
	int lheight,rheight,ldiameter,rdiameter;
	lheight = height(temp->left);
	rheight = height(temp->right);
	dia = 1 + lheight + rheight;
	ldiameter = diameter(temp->left);
	rdiameter = diameter(temp->right);
	maxDia = max(dia,max(ldiameter,rdiameter));
	return maxDia;
}
int main()
{
	int i,an;
	int x[]={41,27,83,45,65,80,40,55};
	/*
		   	 41
		    / \
		   27  83
		   \  /
		   40 45
		      \
		       65
		       / \
		     55  80
	*/
	for(i=0;i<=7;i++)
	newnode(x[i]);
	an=diameter(root);
	cout<<"\nDiameter of tree:"<<an;
	return 0;
}

File: data_structures/dimaeter_of_a_tree/java/Diameter.java
import java.lang.Math;


class Diameter{
	Node root;

	public Diameter(Node root){
		this.root = root;
	}

   public static int c_diameter(Node node){

   	if(node == null)return 0;


    //heights of left & right subtrees
   	int c_left_height = c_height(node.left);
   	int c_right_height = c_height(node.right);

   	//diameters of left an right subtrees
   	int c_left_diameter = c_diameter(node.left);
   	int c_right_diameter = c_diameter(node.right);

    //return 
    return  Math.max(c_left_height+c_right_height+1,Math.max(c_left_diameter,c_right_diameter));

   }


   public static int c_height(Node node){
   	if(node == null) return 0;


   	return 1 + Math.max(c_height(node.left),c_height(node.right));
   }





	public static void main(String[] args){
		Node root = new Node(1);
		root.left = new Node(2);
		root.right = new Node(3);
		root.left.left = new Node(4);
		root.left.right = new Node(5);
		root.left.left.left = new Node(6);

		Diameter diameter = new Diameter(root);

		System.out.println(c_diameter(root));

	}
}